src/coreclr/vm/unsafeaccessors.cpp

```diff
@@ -75,16 +93,249 @@ namespace
 
         UnsafeAccessorKind Kind;
         MethodDesc* Declaration;
-        MetaSig DeclarationSig;
+        SigPointer DeclarationSig;  // This is the official declaration signature. It may be modified
+                                    // to include the UnsafeAccessorTypeAttribute types.
+        MetaSig DeclarationMetaSig;
+        NewArrayHolder<ParamDetails> TranslatedParams; // Redefined types from UnsafeAccessorTypeAttribute usage.
+                                                       // Return type is at 0 index. Function arguments are 1 to N.
         SigPointer TargetTypeSig;
         TypeHandle TargetType;
         bool IsTargetStatic;
         MethodDesc* TargetMethod;
         FieldDesc* TargetField;
     };
 
+    void AppendTypeToSignature(
+        SigBuilder& sig,
+        TypeHandle th,
+        Instantiation inst)
+    {
+        STANDARD_VM_CONTRACT;
+        _ASSERTE(!th.IsNull());
+
+        //
+        // Building the signature follows details defined in ECMA-335 - II.23.2.12
+        //
+
+        // If we have any generic variables, mark as ELEMENT_TYPE_GENERICINST.
+        BOOL hasGenericVariables = inst.ContainsGenericVariables();
+        if (hasGenericVariables)
+            sig.AppendElementType(ELEMENT_TYPE_GENERICINST);
+
+        // Append the new type to the signature.
+        sig.AppendElementType(ELEMENT_TYPE_INTERNAL);
+        sig.AppendPointer(th.AsPtr());
```

Comment:
If we are encoding generic type with variables, we should encode generic type definition here.